src/app/composer/qgscomposermanager.cpp

```diff
@@ -338,101 +389,86 @@ void QgsComposerManager::changeEvent( QEvent* event )
 
-  //ask for confirmation
-  if ( QMessageBox::warning( this, tr( "Remove composer" ), tr( "Do you really want to remove the map composer '%1'?" ).arg( item->text() ), QMessageBox::Ok | QMessageBox::Cancel ) != QMessageBox::Ok )
+  if ( composersItems.isEmpty() )
   {
     return;
   }
 
-  //delete composer
-  QMap<QListWidgetItem*, QgsComposer*>::iterator it = mItemComposerMap.find( item );
-  if ( it != mItemComposerMap.end() )
+  // Ask for confirmation
+  if ( composersItems.count() == 1)
   {
-    QgisApp::instance()->deleteComposer( it.value() );
+    title = tr( "Remove composer" );
+    QListWidgetItem* uniqItem = composersItems.at( 0 );
+    message = tr( "Do you really want to remove the map composer '%1'?" ).arg( uniqItem->text() );
   }
-}
 
-void QgsComposerManager::show_clicked()
-{
-  QListWidgetItem* item = mComposerListWidget->currentItem();
-  if ( !item )
+  if ( QMessageBox::warning( this, title, message, QMessageBox::Ok | QMessageBox::Cancel ) != QMessageBox::Ok )
   {
     return;
   }
 
-  QMap<QListWidgetItem*, QgsComposer*>::iterator it = mItemComposerMap.find( item );
-  if ( it != mItemComposerMap.end() )
+  // Find the QgsComposers that need to be deleted
+  Q_FOREACH ( QListWidgetItem* item, composersItems )
   {
-    QgsComposer* c = 0;
-    if ( it.value() ) //a normal composer
+    QMap<QListWidgetItem*, QgsComposer*>::iterator it = mItemComposerMap.find( item );
+    if ( it != mItemComposerMap.end() )
     {
-      c = it.value();
-      if ( c )
-      {
-        // extra activation steps for Windows
-        bool shown = c->isVisible();
-
-        c->activate();
-
-        // extra activation steps for Windows
-        if ( !shown )
-        {
-          c->on_mActionZoomAll_triggered();
-        }
-      }
+      composersList << it.value();
     }
   }
-#if 0
-  else //create composer from default template
+  
+  // Once we have the composer list, we can delete all of them !
+  Q_FOREACH ( QgsComposer* c, composersList )
   {
-    QMap<QString, QString>::const_iterator templateIt = mDefaultTemplateMap.find( it.key()->text() );
-    if ( templateIt == mDefaultTemplateMap.constEnd() )
-    {
-      return;
-    }
-
-    QDomDocument templateDoc;
-    QFile templateFile( templateIt.value() );
-    if ( !templateFile.open( QIODevice::ReadOnly ) )
-    {
-      return;
-    }
-
-    if ( !templateDoc.setContent( &templateFile, false ) )
-    {
-      return;
-    }
-    c = QgisApp::instance()->createNewComposer();
-    c->setTitle( it.key()->text() );
-    if ( c )
-    {
-      c->readXML( templateDoc );
-    }
+      QgisApp::instance()->deleteComposer( c );
   }
+}
 
-  if ( c )
+void QgsComposerManager::show_clicked()
+{
+  Q_FOREACH ( QListWidgetItem* item, mComposerListWidget->selectedItems() )
   {
-    c->activate();
+    QMap<QListWidgetItem*, QgsComposer*>::iterator it = mItemComposerMap.find( item );
+    if ( it != mItemComposerMap.end() )
```

Comment:
Go:

     QMap<QListWidgetItem*, QgsComposer*>::const_iterator it = mItemComposerMap.constFind( item );
     if ( it != mItemComposerMap.constEnd() ) 

instead here, using the const version is generally faster